### DISCUSSION TITLE
Sealed Secrets image pull issues

### DISCUSSION THREAD

**Discussion by c0c0n3** [c0c0n3/kitt4sme.live#144] [Q&A]
Question originally asked by @etoschi in the [old forum](https://github.com/c0c0n3/kitt4sme/discussions/12).

I noticed the following error ImagePullBackOff when examining the just installed demo cluster :

 
```console
$ kubectl get pods -n kube-system

NAME                                         READY   STATUS             RESTARTS   AGE

hostpath-provisioner-566686b959-h5c2k        1/1     Running            1          25h

calico-kube-controllers-f7868dd95-rj8jv      1/1     Running            1          25h

coredns-7f9c69c78c-xjb52                     1/1     Running            1          25h

calico-node-5ckf2                            1/1     Running            1          25h

reloader-55885ccf44-5xmjl                    1/1     Running            0          25h

sealed-secrets-controller-75445c88df-f4249   0/1     ImagePullBackOff   0          25h

 

kubectl describe pod sealed-secrets-controller-75445c88df-f4249 -n kube-system

 

...

Events:

  Type    Reason   Age                    From     Message

  ----    ------   ----                   ----     -------

  Normal  Pulling  52m (x227 over 19h)    kubelet  Pulling image "[quay.io/bitnami/sealed-secrets-controller:v0.17.3](http://quay.io/bitnami/sealed-secrets-controller:v0.17.3)"

  Normal  BackOff  2m6s (x5246 over 19h)  kubelet  Back-off pulling image "[quay.io/bitnami/sealed-secrets-controller:v0.17.3](http://quay.io/bitnami/sealed-secrets-controller:v0.17.3)"
```

**Comment by c0c0n3** [comment]
Sometimes K8s can get stuck in a pull image retry loop---e.g. the docker registry is unavailable, back-off time growing exponentially, etc.

Can you try deleting the pod? Because the deployment is still there, K8s should try pulling that image again...

**Comment by c0c0n3** [comment]
Reposting @etoschi's reply from the old forum [thread][reply].

I tried to delete the pod as suggested, but still getting ImagePullBackOff – there’s a difference, though, in that now the Events: section in the describe command for the pod provides more output, in particular showing a 401 UNAUTHORIZED message.

Events:

```
  Type     Reason     Age                From               Message

  ----     ------     ----               ----               -------

  Normal   Scheduled  60s                default-scheduler  Successfully assigned kube-system/sealed-secrets-controller-75445c88df-8sq7m to algreenwaste

  Normal   Pulling    15s (x3 over 59s)  kubelet            Pulling image "[quay.io/bitnami/sealed-secrets-controller:v0.17.3](http://quay.io/bitnami/sealed-secrets-controller:v0.17.3)"

  Warning  Failed     13s (x3 over 58s)  kubelet            Failed to pull image "[quay.io/bitnami/sealed-secrets-controller:v0.17.3](http://quay.io/bitnami/sealed-secrets-controller:v0.17.3)": rpc error: code = Unknown desc = failed to pull and unpack image "[quay.io/bitnami/sealed-secrets-controller:v0.17.3](http://quay.io/bitnami/sealed-secrets-controller:v0.17.3)": failed to resolve reference "[quay.io/bitnami/sealed-secrets-controller:v0.17.3](http://quay.io/bitnami/sealed-secrets-controller:v0.17.3)": unexpected status code [manifests v0.17.3]: 401 UNAUTHORIZED

  Warning  Failed     13s (x3 over 58s)  kubelet            Error: ErrImagePull

  Normal   BackOff    1s (x3 over 58s)   kubelet            Back-off pulling image "[quay.io/bitnami/sealed-secrets-controller:v0.17.3](http://quay.io/bitnami/sealed-secrets-controller:v0.17.3)"

  Warning  Failed     1s (x3 over 58s)   kubelet            Error: ImagePullBackOff
```

I tried several times to delete the pod, just to be sure, but always getting the same result.



[reply]: https://github.com/c0c0n3/kitt4sme/discussions/12#discussioncomment-2688275

**Comment by c0c0n3** [comment]
@etoschi there's a problem with that image at the moment. We're busy developing a hot-fix you should be able to apply next week. You can track progress over here
-  #147
- https://github.com/users/c0c0n3/projects/1/views/20

**Comment by c0c0n3** [comment]
@etoschi @cmelchiorre we've just merged a hotfix (#148) into the mainline which should work out this issue. if you rebase your repo you should be able to use sealed secrets again. sorry it took a long time to get done.

**Comment by cmelchiorre** [comment]
Hi @c0c0n3,

I did the rebase – strange thing, when I try to reconnect to my server to check everything is now ok, if I launch `~/tools.sh` it starts downloading everything again (which it didn’t do usually after the first installation...); after a while it stops with the error shown below.
This doesn’t look like something related to the changes made during the rebase, but rather some problem in connecting with the server proxy.golang.org, but I cannot be sure of this.

Do you have any idea whether this could related to the changes made? 

By the way, I could check from a kubectl on my pc configured to connect to the cluster that the pods for sealed-secrets-controller are working.

`  error: builder for '/nix/store/4vzgbmn0mqqadi7swf278qdj97y26ysn-kubeseal-0.17.5.drv' failed with exit code 1;
         last 10 log lines:
         > pkg/apis/sealed-secrets/v1alpha1/sealedsecret_expansion.go:14:2: k8s.io/apimachinery@v0.23.4: Get "https://proxy.golang.org/k8s.io/apimachinery/@v/v0.23.4.zip": dial tcp: lookup proxy.golang.org on [::1]:53: read udp [::1]:56328->[::1]:53: read: connection refused
         > cmd/kubeseal/main.go:28:2: k8s.io/apimachinery@v0.23.4: Get "https://proxy.golang.org/k8s.io/apimachinery/@v/v0.23.4.zip": dial tcp: lookup proxy.golang.org on [::1]:53: read udp [::1]:56328->[::1]:53: read: connection refused
         > pkg/apis/sealed-secrets/v1alpha1/register.go:7:2: k8s.io/apimachinery@v0.23.4: Get "https://proxy.golang.org/k8s.io/apimachinery/@v/v0.23.4.zip": dial tcp: lookup proxy.golang.org on [::1]:53: read udp [::1]:56328->[::1]:53: read: connection refused
         > pkg/apis/sealed-secrets/v1alpha1/register.go:8:2: k8s.io/client-go@v0.23.4: Get "https://proxy.golang.org/k8s.io/client-go/@v/v0.23.4.zip": dial tcp: lookup proxy.golang.org on [::1]:53: read udp [::1]:42144->[::1]:53: read: connection refused
         > cmd/kubeseal/main.go:30:2: k8s.io/client-go@v0.23.4: Get "https://proxy.golang.org/k8s.io/client-go/@v/v0.23.4.zip": dial tcp: lookup proxy.golang.org on [::1]:53: read udp [::1]:42144->[::1]:53: read: connection refused
         > cmd/kubeseal/main.go:43:2: k8s.io/client-go@v0.23.4: Get "https://proxy.golang.org/k8s.io/client-go/@v/v0.23.4.zip": dial tcp: lookup proxy.golang.org on [::1]:53: read udp [::1]:42144->[::1]:53: read: connection refused
         > cmd/kubeseal/main.go:31:2: k8s.io/client-go@v0.23.4: Get "https://proxy.golang.org/k8s.io/client-go/@v/v0.23.4.zip": dial tcp: lookup proxy.golang.org on [::1]:53: read udp [::1]:42144->[::1]:53: read: connection refused
         > cmd/kubeseal/main.go:32:2: k8s.io/client-go@v0.23.4: Get "https://proxy.golang.org/k8s.io/client-go/@v/v0.23.4.zip": dial tcp: lookup proxy.golang.org on [::1]:53: read udp [::1]:42144->[::1]:53: read: connection refused
         > cmd/kubeseal/main.go:33:2: k8s.io/client-go@v0.23.4: Get "https://proxy.golang.org/k8s.io/client-go/@v/v0.23.4.zip": dial tcp: lookup proxy.golang.org on [::1]:53: read udp [::1]:42144->[::1]:53: read: connection refused
         > cmd/kubeseal/main.go:34:2: k8s.io/klog/v2@v2.40.1: Get "https://proxy.golang.org/k8s.io/klog/v2/@v/v2.40.1.zip": dial tcp: lookup proxy.golang.org on [::1]:53: read udp [::1]:42377->[::1]:53: read: connection refused
         For full logs, run 'nix log /nix/store/4vzgbmn0mqqadi7swf278qdj97y26ysn-kubeseal-0.17.5.drv'.
  error: 1 dependencies of derivation '/nix/store/gcxa0icdrv6d5zy9qh84zlyrx8hi1vz9-kitt4sme-cluster-shell.drv' failed to build
`

**Comment by c0c0n3** [comment]
Hi @cmelchiorre :-)

>  ~/tools.sh it starts downloading everything again

It's going to build the new Sealed Secrets client---`kubeseal` command. I had to upgrade it (#148) to version `0.17.5` to make sure the client version is the same as the new service version deployed in the cloud. Your Nix store had the previous version (`0.17.3`) so it's now trying to build `0.17.5` but to do that, it needs to download the Go source code first and then all the dependent libs.

> rather some problem in connecting with the server proxy.golang.org

Yep, that's exactly the reason---well at least judging by the error trace you copy-pasted above. Are you behind a firewall? i.e. is the box where you run `tools.sh` behind a firewall that doesn't let through traffic to port `53`? If traffic to port `53` is allowed, then just try running `tools.sh` again, it could be the Go proxy site was overloaded and couldn't accept any more connections, so if you run the Nix build (`tools.sh` triggers it) a few more times it should eventually be able to download all the deps and then finally build the `kubeseal` executable.

> related to the changes made?

Yes, my changes, not yours :-) Like I said earlier, I had to update the dev tools.

> I could check from a kubectl on my pc configured to connect to the cluster that the pods for sealed-secrets-controller are working.

Excellent! The problem you're having has nothing to do with the cluster. It's just the dev tools we use to connect to the cluster that are affected.

**Comment by c0c0n3** [comment]
@cmelchiorre we've just merged a fix for the Nix connection error issue, see #171.
